doc/dev/code-generator.md

```diff
@@ -0,0 +1,211 @@
+  string `""` is a valid part suffix, although it is only recommend for templates with a single output file.
+
+## Creating a new template
+
+In this guide we will create a basic template, that generates a single file containing. The file will simple list all keys in our input
+keyset. An example output would be (the same as `kdb ls`):
+
+```
+user/sw/myapp/#0/current
+user/sw/myapp/#0/current/dir0
+user/sw/myapp/#0/current/dir0/subdir0/key0
+user/sw/myapp/#0/current/dir0/subdir0/key1
+user/sw/myapp/#0/current/dir1
+```
+
+### Creating the mustache template
+
+The first thing we need to create is a mustache template. You will need (at least) one template file for each output file. If you make use
+of partials (see [below](#using-partials)), you of course have to write multiple files.
+
+All template files must be located in the [`src/tools/kdb/gen/templates`](/src/tools/kdb/gen/templates) folder and must be named according
+to the scheme `<template base name>.<part suffix>.mustache`.
+
+For our simple example we create the file `src/tools/kdb/gen/templates/example.txt.mustache` with the following content:
+
+```
+{{# keys }}
+{{{ name }}}
+{{/ keys }}
+```
+
+Note: we will not go into detail on how mustache templates work, for more information see e.g.
+[here](https://mustache.github.io/mustache.5.html). All of features supported by the kainjow library should be supported by our framework
+as well.
+
+### Creating the supporting class
+
+Since we need to some way of supplying data to our template, we have to create a subclass of `GenTemplate`.
+
+First we create a new directory for our template class in `src/tools/kdb/gen`, for our template it should be `src/tools/kdb/gen/example`. In
+this directory we then create `example.cpp` and `example.hpp`. If your template becomes sufficiently complex, it may make sense to split it
+into multiple files, for this reason we recommend creating a new directory for each template.
+
+In `example.cpp` and `example.hpp` we create our subclass of `GenTemplate`. Therefore `example.hpp` should look like this:
+
+```cpp
+#ifndef ELEKTRA_EXAMPLE_HPP // choose a unique header guard
+#define ELEKTRA_EXAMPLE_HPP
+
+#include "../template.hpp"
+
+class ExampleGenTemplate : public GenTemplate
+{
+public:
+    ExampleGenTemplate () : GenTemplate ("example", { ".txt" }, {}, {})
+    {
+    }
+
+protected:
+    kainjow::mustache::data getTemplateData (const std::string & outputName, const std::string & part, const kdb::KeySet & ks,
+                                             const std::string & parentKey) const override;
+};
+
+#endif // ELEKTRA_EXAMPLE_HPP
+```
+
+Apart from the line `ExampleGenTemplate () : GenTemplate ("example", { ".txt" }, {}, {})` everything should be more or less the same for
+all templates. Let's dissect this line:
+
+- `ExampleGenTemplate ()` we declare a zero argument constructor. All templates _must_ have only a single zero argument constructor,
+  otherwise the framework cannot instantiate them.
```

Comment:
Which framework?

Reply:
The code-generator framework. I called the internal setup that invokes the mustache library "framework" to distinguish it from the command line interface implemented in `gen.cpp`.

Reply:
Then describe this when introducing the terms.

Reply:
It is actually explained in this paragraph at the start of the Basics section:

> The file [`src/tools/kdb/gen.cpp`](/src/tools/kdb/gen.cpp) implements the command-line interface (CLI) and is of little interest to this guide. Instead we will focus on the framework that is invoked via the CLI. The bulk of the framework is implemented in the classes [`GenTemplate` and `GenTemplateList`](/src/tools/kdb/gen/template.hpp) (and of course kainjow's mustache library).

I will add further clarification.